misc: introduce `uv_get|set_threadpool_size`

juanarbol opened the pull request:
This call will set the threadpool size from libuv and set `UV_THREADPOOL_SIZE` environment variable.

Fixes: https://github.com/libuv/libuv/issues/4401


----
This is my approach for https://github.com/libuv/libuv/issues/4401; and of course I'm open to changes and suggestions.
I decided to *mutate* `UV_THREADPOOL_SIZE` considering that maybe some Node.js users or anyone relies on that variable for certain situations (like monitoring or something). And this is not a thread safe call as `uv_os_setenv` is not. Even tho, I don't is possible make this thread safe on any way.

cjihrig commented:
`uv_set_threadpool_size()` seems like a more intuitive name IMO.

juanarbol commented:
> `uv_set_threadpool_size()` seems like a more intuitive name IMO.

I think you are right. Will rename and fix the windows under ASAN. It seems to be related with https://github.com/libuv/libuv/issues/4338

saghul commented on docs/src/misc.rst:
```diff
@@ -886,3 +886,15 @@ is not complete.
+    This function must be called before any other libuv function is called.
+
+    On success, it returns 0, if the threadpool is already spining it returns
+    `UV_EINVAL`.
```
Maybe UV_EALREADY or UV_EBUSY are better error codes?

saghul commented on src/threadpool.c:
```diff
@@ -261,6 +261,26 @@ static void init_once(void) {
+  char num_str[12]; /* enough room for unsigned int and null terminator */
+
+  /* check if the threadpool is already initialized */
+  if (nthreads != 0)
```
I wonder: would it be possible to enlarge the amount of threads while the pool is running?

Reducing it is probably hard / impossible, but adding new threads might be within the realm of possibility.

juanarbol replied:
I think that is out of the scope of the issue itself. I could work on that as well. It could be a mix between @santigimeno suggestion of a getter for this.

bnoordhuis replied:
I have a branch from 2015 that does just that: https://github.com/bnoordhuis/libuv/commits/auto-threadpool

Shrinking is done lazily. Threads terminate when they're idle and the actual thread count > desired thread count.

juanarbol replied:
@bnoordhuis  Do you think I should pivot this approach into something like yours instead?

bnoordhuis replied:
I'd say that's a good idea. https://github.com/libuv/libuv/pull/382 is the PR from back then, you will want to read through the discussion.

cc @saghul as a fyi. You had some comments back then; maybe you still do?

joy4eg commented on docs/src/misc.rst:
```diff
@@ -886,3 +886,15 @@ is not complete.
 
     .. versionadded:: 1.47.0
+
+.. c:function:: int uv_set_threadpool_size(unsingned int number)
```
Should we use `size_t` ?

juanarbol replied:
Makes sense

juanarbol replied:
Nevermind, the code seems to like unsigned int instead.

saghul replied:
Your spin_threads function takes size_t

santigimeno commented on include/uv.h:
```diff
@@ -1313,6 +1313,8 @@ enum {
 UV_EXTERN int uv_thread_getpriority(uv_thread_t tid, int* priority);
 UV_EXTERN int uv_thread_setpriority(uv_thread_t tid, int priority);
 
+UV_EXTERN int uv_set_threadpool_size(unsigned int number);
+
```
Adding a `uv_get_threadpool_size()` would be a good idea, I think.

juanarbol replied:
That could be useful as well

santigimeno commented on src/threadpool.c:
```diff
@@ -261,6 +261,26 @@ static void init_once(void) {
+  if (nthreads != 0)
+    return UV_EINVAL;
+
+  /* Set the new UV_THREADPOOL_SIZE */
+  sprintf(num_str, "%u", number);
+  r = uv_os_setenv("UV_THREADPOOL_SIZE", num_str);
+  if (r != 0)
+    return r;
```
I'm not fully convinced about the "users are going to be checking the UV_THREADPOOL_SIZE env variable" argument. I'd rather add a function to get the number of threads as mentioned before. Also how is this going to work? I guess in order not to break anything `UV_THREADPOOL_SIZE` should take precedence over the API call? I don't know tbh.

juanarbol replied:
The nthreads variable will hold the value of the current number of threads. That is what I saw during development of this.

This is not breaking anything -as it is rn-, I considered mutating the env as a side effect as the easiest way to solve this and have predictable output. I’m not sure about having a different value for that env var and having a different number of available threads in the thread pool

saghul replied:
But that is always the case if the user sets the value in the env variable too late...

IMHO I wouldn't set the env variable. Just take the passed value and use it. The better @santigimeno proposed serves as a way to read the actual value.

Note there is another thing, IIRC we cap the value we read from the env var, so there is already precedent on the env var and reality being out of sync. We should do the same here, apply a cap, and not set the env var.

juanarbol commented:
Thanks for the review, everyone, after reading all the comments and suggestions.

1. I will extend the functionalities over the setter, if `UV_THREADPOOL_SIZE` is not provided, will spin as many threads as requested in this call. If env is provided; that will be rule and there is nothing users can do to spin more threads or a different number of threads.
2. If called again with more threads, spins k more threads
3. Introduce the getter.

juanarbol commented:
The new revision includes the getter and the behaviour is a bit different now as this patch surpasses the functionalities of the original issue. And most importantly, it respects `UV_THREADPOOL_SIZE` as the first source of truth and won't mutate it.

juanarbol commented:
I'll see what is wrong with Windows later, maybe tomorrow, but feel free to review anyway. </3

juanarbol commented:
```c
r = uv_os_setenv("UV_THREADPOOL_SIZE", "1");
ASSERT_OK(r);
#ifdef _WIN32
  /* juanarbol: why? */
  ASSERT_EQ(nthreads, 4);
#else
  /* The default thread pool size is UV_THREADPOOL_SIZE */
  ASSERT_EQ(nthreads, 1);
#endif
```

This is odd. If anyone can confirm if windows simply does not work or have any idea about this, would be really appreciated.

saghul commented on src/threadpool.c:
```diff
@@ -190,12 +190,31 @@ void uv__threadpool_cleanup(void) {
 
 
-static void init_threads(void) {
+void spin_threads(size_t from, size_t to) {
```
Make this static

saghul commented on src/threadpool.c:
```diff
@@ -206,7 +225,7 @@ static void init_threads(void) {
 
   threads = default_threads;
-  if (nthreads > ARRAY_SIZE(default_threads)) {
+  if (nthreads != ARRAY_SIZE(default_threads)) {
```
Why allocate memory if we need fewer threads than the default?

juanarbol replied:
Yeah. Right, that was from a different experiment

saghul commented on src/threadpool.c:
```diff
@@ -261,6 +268,62 @@ static void init_once(void) {
+
+  /* Ensure the threadpool is initialized */
+  if (nthreads == 0)
+    uv_once(&once, init_once);
```
Since it's a once handle you can call it every time.

juanarbol replied:
With that one. Idk how’s gonna init the thread pool first. I just  used uv_once on all possibilities -just in case-

saghul replied:
You actually have to call it first and foremost, then get the mutex, then you can check the value of `nthreads`. It's not safe otherwise.

saghul commented on src/threadpool.c:
```diff
@@ -261,6 +268,62 @@ static void init_once(void) {
+    uv_once(&once, init_once);
+
+  /* no-op if the threadpool size is already set */
+  if (number == nthreads)
```
What if 2 threads are calling this function at the same time? You might need a mutex.

juanarbol replied:
I wasn’t considering this to be thread safe 🫠🥲 but yeah, I can work on that

saghul replied:
Yeah you do need the mutex since you are mutating the global state.

saghul commented on src/threadpool.c:
```diff
@@ -261,6 +268,62 @@ static void init_once(void) {
+  }
+
+  /* Alloc -with the expected size- if was not alloc(ed) before */
+  if (threads == default_threads) {
```
I need to check but I think this will need a lock around it...

juanarbol replied:
Is it possible that the tear down call needs a mutex as well?

saghul replied:
This operation needs it, not the teardown since it happens at library destruction time.

saghul commented on include/uv.h:
```diff
@@ -1313,6 +1313,9 @@ enum {
 UV_EXTERN int uv_thread_setpriority(uv_thread_t tid, int priority);
 
+UV_EXTERN int uv_set_threadpool_size(unsigned int number);
+UV_EXTERN int uv_get_threadpool_size(unsigned int* number);
```
Why not simply return the value, rather than getting an out argument?

juanarbol replied:
Passing an arg pointer and fill its value is the most common pattern (IMO) in libuv, all calls seems to return errno like instead of values (that’s what I normally see). I can modify it for your suggestion, but I personally think that breaks the most common pattern here

saghul replied:
```
UV_EXTERN size_t uv_loop_size(void);
UV_EXTERN size_t uv_handle_size(uv_handle_type type);
...

UV_EXTERN int uv_send_buffer_size(uv_handle_t* handle, int* value);
UV_EXTERN int uv_recv_buffer_size(uv_handle_t* handle, int* value);

UV_EXTERN size_t uv_stream_get_write_queue_size(const uv_stream_t* stream);
UV_EXTERN size_t uv_udp_get_send_queue_size(const uv_udp_t* handle);
UV_EXTERN int uv_cpumask_size(void);

```

For size shenanigans, looks like just returning the size is the more used pattern.

juanarbol replied:
Yeap, will update that one.

joy4eg commented:
> ```c
> r = uv_os_setenv("UV_THREADPOOL_SIZE", "1");
> ASSERT_OK(r);
> #ifdef _WIN32
>   /* juanarbol: why? */
>   ASSERT_EQ(nthreads, 4);
> #else
>   /* The default thread pool size is UV_THREADPOOL_SIZE */
>   ASSERT_EQ(nthreads, 1);
> #endif
> ```
> 
> This is odd. If anyone can confirm if windows simply does not work or have any idea about this, would be really appreciated.

setenv() just sets the env, nothing more. 4 is the default thread pool size, and that's why you got it.

juanarbol reviewed (commented):
Thanks! I will address all the comments and wait for your opinions if we need all those locks

juanarbol commented:
> ```c
> ```c
> r = uv_os_setenv("UV_THREADPOOL_SIZE", "1");
> ASSERT_OK(r);
> #ifdef _WIN32
>   /* juanarbol: why? */
>   ASSERT_EQ(nthreads, 4);
> #else
>   /* The default thread pool size is UV_THREADPOOL_SIZE */
>   ASSERT_EQ(nthreads, 1);
> #endif
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> This is odd. If anyone can confirm if windows simply does not work or have any idea about this, would be really appreciated.
> 
> setenv() just sets the env, nothing more. 4 is the default thread pool size, and that's why you got it.
> ```

I was expecting that env var to affect the threadpool size, like it does in XNIX

saghul commented on src/threadpool.c:
```diff
@@ -224,20 +243,8 @@ static void init_threads(void) {
-
-  uv_sem_destroy(&sem);
+  /* Spin up the default number of threads */
+  spin_threads(0 /* from */, nthreads /* to */);
```
This call is fine because it's inside `init_threads`, which is guarateed to only run once. Any other place where this is called needs to hold onto the mutex.

saghul commented on src/threadpool.c:
```diff
@@ -261,6 +268,62 @@ static void init_once(void) {
+int uv_get_threadpool_size(unsigned int* number) {
+  /* check if the threadpool is already initialized first */
+  if (nthreads == 0)
+    uv_once(&once, init_once);
```
Ditto, call it first, no checking.

saghul commented on test/test-threadpool-size.c:
```diff
@@ -0,0 +1,99 @@
+  RETURN_SKIP("Test does not currently work on Windows under ASAN");
+#endif
+
+  r = uv_os_setenv("UV_THREADPOOL_SIZE", "1");
```
Not sure how reliable this is, since any chance in the test runner what uses any fs API will trigger a threadpool initialization.

juanarbol replied:
Is not each test -at least in this case- like a "isolated" event loop?

juanarbol commented:
Ok, after addressing all the comments here, I think this is thread safe now :)

Thanks for reviewing, explanations and your time. Of course I'm open for more changes if needed.

juanarbol commented:
After commenting the the `uv__io_stop` and arm that again. I'm getting some failures in `delayed_accept`; I guess it just makes sure that this EMFILE magic trick is always gonna work. See:

```
/* Implements a best effort approach to mitigating accept() EMFILE errors.
 * We have a spare file descriptor stashed away that we close to get below
 * the EMFILE limit. Next, we accept all pending connections and close them
 * immediately to signal the clients that we're overloaded - and we are, but
 * we still keep on trucking.
 *
 * There is one caveat: it's not reliable in a multi-threaded environment.
 * The file descriptor limit is per process. Our party trick fails if another
 * thread opens a file or creates a socket in the time window between us
 * calling close() and accept().
 */
```

https://github.com/libuv/libuv/blob/d2d92b74a8327daf9652a9732454937f3529bd30/src/unix/stream.c#L480

```
not ok 1 - delayed_accept
# exit code 6
# Output from process `delayed_accept`:
# Assertion failed in /Users/juanjose/GitHub/libuv/test/test-delayed-accept.c on line 61: `r == 0` (-35 == 0)
```

IMO the IO handle stop/start-again mechanism works like a lock to avoid other threads read a file or create a socket.

juanarbol commented:
Humble ping @saghul, the CI failure is unrelated to this. And about the environment test case, I think it is worth having it, that describes the whole feature behaviour. But I can remove it if needed.

juanarbol commented:
@bnoordhuis I've been bouncing my head on this one for a couple weeks now.

I see some problems having an "elastic _threadpool_", your approach on https://github.com/libuv/libuv/pull/382 has a very specific flaw when no threads are available but the work queue still not empty [1].

For that, I considered another thread to manage this new "detached-threadpool", that will slurp from the queues and spawn threads when possible (libuv has not spawned all the _configured_ threads). But that is basically adding un-necessary complexity.

On the other hand, I think a threadpool (as is, keep the threads alive) seems to be more efficient/performant than creating threads on demand to run the so-called "work", and that is being pointed by your comment here https://github.com/libuv/libuv/pull/382/files#diff-9f73dfa55c121c137b316c5b5f1fc25f0384b156add5f3e1c61f396311b568f1R215

I'm open to receive some guidance on next steps for this PR. The get/set threadpool-size seems like a nice feature.

[1]: https://github.com/libuv/libuv/pull/382/files#diff-9f73dfa55c121c137b316c5b5f1fc25f0384b156add5f3e1c61f396311b568f1R235

juanarbol commented:
Note aside, we could make an elastic threadpool if we can identify when a thread is idling and get rid of it/them (the setter is requesting a "shrink") using the current approach of storing them in an array. But not very sure how can we mark a thread as idling.

bnoordhuis commented:
> I considered another thread

That won't work because that code path you link to is specifically for when threads can't be started because we're at the per-user or system limit.

I also wouldn't call it a flaw because right now libuv just aborts when threads can't be started.

juanarbol commented:
> That won't work because that code path you link to is specifically for when threads can't be started because we're at the per-user or system limit.
> 
> I also wouldn't call it a flaw because right now libuv just aborts when threads can't be started.

Got it; any suggestions? Ideas that could lead me into a decent solution?

bnoordhuis commented:
If we're talking specifically about the inability to start threads, there are only 3 possible solutions:

1. do the work synchronously
2. return an error
3. abort

(Option 4, quietly ignore and drop the work, is not really an option.)

juanarbol commented:
@bnoordhuis I think this new revision has a better shape, I'll have to add docs for it again, as the behavior changed a bit.

Please let me know any comments or ideas for this one. For know, the unit test is passing consistently. I've tested only TSAN, for now, as this is meant to be thread safe. If any leaks are present (or any kind of failure) and this is closer to its final shape, I'll fix them.

juanarbol commented:
Valgrind seems happy with helgrind and memory leaks.

```sh
$ repeat 1000 ./build/uv_run_tests_a threadpool_size threadpool_size

$ valgrind \
--tool=helgrind \
./build/uv_run_tests_a threadpool_size threadpool_size
...
==65456== For lists of detected and suppressed errors, rerun with: -s
==65456== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 62 from 23)

$ valgrind \
--leak-check=full \
--show-leak-kinds=all \
--track-origins=yes \
--verbose \
./build/uv_run_tests_a threadpool_size threadpool_size
...
==66623== HEAP SUMMARY:
==66623==     in use at exit: 0 bytes in 0 blocks
==66623==   total heap usage: 9 allocs, 9 frees, 2,063 bytes allocated
==66623==
==66623== All heap blocks were freed -- no leaks are possible
==66623==
==66623== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

I've introduced https://github.com/libuv/libuv/pull/4415/commits/b856f4b2940b9abf234bd000cb17ef5fdc5e523e trying to hunt the timeouts. As long as we can just "join" _idle_ threads in the threadpool, it may take some time while another thread holds the lock or something. But with a 1k times test consistency, I genuinely believe this patch behaves as it should.

Any recommendations?
cc @saghul @bnoordhuis @santigimeno

santigimeno commented on test/test-threadpool-size.c:
```diff
@@ -0,0 +1,47 @@
+  ASSERT_EQ(uv_get_threadpool_size(), 5);
+
+  /* Shrink the thread pool */
+  /*
+  ASSERT_OK(uv_set_threadpool_size(4));
+  ASSERT_EQ(uv_get_threadpool_size(), 4);
+  */
```
I guess this should be uncomented right?

juanarbol replied:
Right. I just commented due to -mostly- qemu timeouts. without that the CI is very red.

santigimeno replied:
Those timeouts are strange. Can you increase the timeout for that specific test and push the changes just to be sure we're not missing smthg?

santigimeno commented on src/threadpool.c:
```diff
@@ -262,6 +279,143 @@ static void init_once(void) {
+  if (r)
+    goto out;
+
+  while (nthreads < n) {
+    /* Wire up the error code to the return value. */
+    r = uv_thread_create_ex(threads + nthreads, &config, worker, &sem);
+    if (r)
+      break;
+
+    uv_sem_wait(&sem);
+    nthreads++;
+  }
+
+  assert(nthreads == n);
```
In case thread creation fails, the assertion would not hold, right? Probably fixed by replacing the `break` with a `goto out`.

juanarbol replied:
Right. Let me fix that one

santigimeno commented:
BTW, the approach looks good to me.

juanarbol commented:
Apparently windows is not happy with my logic.

```
not ok 381 - threadpool_size
# exit code 1
# Output from process `threadpool_size`:
# =================================================================
# ==4884==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x128c55ca24c0 at pc 0x7ff6246fe95e bp 0x00f6c80ff750 sp 0x00f6c80ff758
# WRITE of size 8 at 0x128c55ca24c0 thread T0
#     #0 0x7ff6246fe95d in uv__threads_join D:\a\libuv\libuv\src\threadpool.c:360
#     #1 0x7ff6246ff4b5 in uv_set_threadpool_size D:\a\libuv\libuv\src\threadpool.c:405
#     #2 0x7ff62467d12d in run_test_threadpool_size D:\a\libuv\libuv\test\test-threadpool-size.c:34
#     #3 0x7ff624596e03 in run_test_part D:\a\libuv\libuv\test\runner.c:372
#     #4 0x7ff624594cfd in main D:\a\libuv\libuv\test\run-tests.c:82
#     #5 0x7ff6247079d3 in invoke_main D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78
#     #6 0x7ff6247079d3 in __scrt_common_main_seh D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
#     #7 0x7ff874484caf  (C:\Windows\System32\KERNEL32.DLL+0x180014caf)
#     #8 0x7ff8754fedca  (C:\Windows\SYSTEM32\ntdll.dll+0x18007edca)
# 
# 0x128c
# 55ca24c0 is located 0 bytes after 32-byte region [0x128c55ca24a0,0x128c55ca24c0)
# allocated by thread T0 here:
#     #0 0x7ff6246b6ba5 in malloc D:\a\_work\1\s\src\vctools\asan\llvm\compiler-rt\lib\asan\asan_malloc_win_thunk.cpp:98
#     #1 0x7ff6246bbdee in uv__malloc D:\a\libuv\libuv\src\uv-common.c:77
#     #2 0x7ff6246fe8c9 in uv__threads_join D:\a\libuv\libuv\src\threadpool.c:351
#     #3 0x7ff6246ff4b5 in uv_set_threadpool_size D:\a\libuv\libuv\src\threadpool.c:405
#     #4 0x7ff62467d12d in run_test_threadpool_size D:\a\libuv\libuv\test\test-threadpool-size.c:34
#     #5 0x7ff624596e03 in run_test_part D:\a\libuv\libuv\test\runner.c:372
#     #6 0x7ff624594cfd in main D:\a\libuv\libuv\test\run-tests.c:82
#     #7 0x7ff6247079d3 in invoke_main D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78
#     #8 0x7ff6247079d3 in __scrt_common_main_seh D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
#     #9 0x7ff874484caf  (C:\Windows\System32\KERNEL32.DLL+0x180014caf)
#     #10 0x7ff8754fed
# ca  (C:\Windows\SYSTEM32\ntdll.dll+0x18007edca)
# 
# SUMMARY: AddressSanitizer: heap-buffer-overflow D:\a\libuv\libuv\src\threadpool.c:360 in uv__threads_join
# Shadow bytes around the buggy address:
#   0x128c55ca2200: 00 00 00 05 fa fa 00 00 02 fa fa fa 00 00 07 fa
#   0x128c55ca2280: fa fa 00 00 00 04 fa fa 00 00 01 fa fa fa 00 00
#   0x128c55ca2300: 06 fa fa fa 00 00 00 01 fa fa 00 00 05 fa fa fa
#   0x128c55ca2380: 00 00 02 fa fa fa fd fd fd fa fa fa fd fd fd fa
#   0x128c55ca2400: fa fa fd fd fd fa fa fa fd fd fd fa fa fa fd fd
# =>0x128c55ca2[480](https://github.com/libuv/libuv/actions/runs/12730135156/job/35482741725#step:7:481): fd fa fa fa 00 00 00 00[fa]fa fd fd fd fd fa fa
#   0x128c55ca2500: fd fd fd fd fa fa fd fd fd fd fa fa fa fa fa fa
#   0x128c55ca2580: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
#   0x128c55ca2600: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
#   0x128c55ca2680: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
#   0x128c55ca2700: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
# Shadow byte legend (one shadow byte represents 8 application bytes):
#   Addressable:           00
#   Parti
# ally addressable: 01 02 03 04 05 06 07 
#   Heap left redzone:       fa
#   Freed heap region:       fd
#   Stack left redzone:      f1
#   Stack mid redzone:       f2
#   Stack right redzone:     f3
#   Stack after return:      f5
#   Stack use after scope:   f8
#   Global redzone:          f9
#   Global init order:       f6
#   Poisoned by user:        f7
#   Container overflow:      fc
#   Array cookie:            ac
#   Intra object redzone:    bb
#   ASan internal:           fe
#   Left alloca redzone:     ca
#   Right alloca redzone:    cb
# ==[488](https://github.com/libuv/libuv/actions/runs/12730135156/job/35482741725#step:7:489)4==ABORTING
```

Why is ASAN not complaining in UNIX? I'm gonna keep working on this one

bnoordhuis reviewed (commented):
Some high-level observations:

1.  I don't like the amount of work this PR does inside the critical section, that's a potential performance bottleneck.

Specifically, allocating memory under lock is something to avoid. The max thread count is capped at 1,024. You can just use a static array.

2. There's quite a bit of song-and-dance going on in order to uv_thread_join the exiting thread. Why not simply detach them?

juanarbol commented:
> allocating memory under lock is something to avoid

Had no idea. I just wanted to make it thread safe.

> The max thread count is capped at 1,024. You can just use a static array.

How can I keep track of each one of them (are they joined, how many are alive...)? I think it makes sense, a detach and a kill_message in the queue.

I know this approach has many many flaws, could you please provide a bit more details on what would you prefer? I can pivot this into something else,  I don't see any problem at all. Thanks.

juanarbol commented:
@bnoordhuis Thanks for the guidance and persistence on this one. Ok, let's use a static array, now, how is the number of threads a consistent number? Detaching the threads won't automagically "join", they will once they are idling or/and a kill_message is in queue. The current patch does have consistency, but jumping into a detaching one, not sure (I don't see it too clear), it's more like a request to join.

bnoordhuis commented:
> how is the number of threads a consistent number?

I'm not exactly sure what you're asking here.

> Detaching the threads won't automagically "join"

That's basically what detach is though, a magic auto-join when the thread exits. With detached threads, you only have to send a "kill yourself" message, no need to call uv_thread_join() and wait until the thread actually exits.

juanarbol commented:
> I'm not exactly sure what you're asking here.

This patch has a getter for the threadpool size, with pthread_detach I won't know -idk how- when a thread is gone and decrease that for that getter, the increment is easier with a semaphore.

bnoordhuis commented:
When the target thread wakes up on receiving the kill message, it has the mutex, so you can just decrement a global counter, as long as you also grab the mutex when reading the counter.

But, taking a step back, what's the point of the getter? The value it returns is, by definition, a snapshot in time. The count may already be out of date by the time it's returned to the caller.

I'd remove it, I don't think it serves a purpose.
